exercise/exercise_summary.py

```diff
@@ -143,6 +145,28 @@ def __init__(self, course_instance):
         self.__collect_student_grades()
 
 
+    def __get_exercises(self):
+        content = CachedContent(self.course_instance)
+
+        def get_descendant_ids(node):
+            children = node['children']
+            if children:
+                return itertools.chain.from_iterable(
+                    [get_descendant_ids(child) for child in children])
+            return (node['id'],)
+
+        root_node = { 'children': content.modules() }
+        ids = get_descendant_ids(root_node)
+
+        # Loop until end of ids raises StopIteration
+        while True:
+            id = next(ids)
+            try:
+                yield LearningObject.objects.get(pk=id).baseexercise
```

Comment:
Could this resolved multiple Learning objects with a single sql query? Or at least base exercises with get(something_learninh_object_id=id). Pretty sure the sql queries can be optimized a bit someway.

Reply:
You're right, this could be done better with `BaseExercise.objects.get(learningobject_ptr_id=id)`.

In principle, it would be possible to hit the database only once by first doing something like `BaseExercise.objects.filter(learningobject_ptr_id__in=ids)` to get the BaseExercises in essentially random order, and then sort them so that the `learningobject_ptr_id`s are in the order they are in `ids`. However, first of all it is not obvious to me that this is more efficient, since the sort could end up becoming a relatively heavy operation. Second, it would be a _much_ more complex solution, and therefore a premature optimisation: the all results page is infrequently used and it takes me about 2500 ms to render the page for O1, so there doesn't seem to be a performance problem.

Reply:
Sounds like the first part (directly requesting baseexercises) is good enough solution.